### PR TITLE
Tutorial about serverless functions

### LINKED ISSUE #2185: How to upload DL which built by myself?
As per the DL models already in the CVAT platform are not good enough, I used more tranning data to built a new model based on YOLOV3, and trained a DL model used for cuboid semi annotation, but I didn't find a way to upload the model, may you share a guideline how to do it?

### LINKED ISSUE #2259: How to use Nuclio once I have deployed as plugin in CVAT?
Hi, I have deployed CVAT with nuclio as a plugin. Now how do I deploy dextr, f-brs, and my own models like maskrcnn and yolov5? I don't think any documentation exists for this yet?

image

I try to run nuctl commands in the nuclio container but I get an error
image

Also, this documentation seems outdated already ? https://github.com/openvinotoolkit/cvat/blob/develop/cvat/apps/documentation/installation.md#semi-automatic-and-automatic-annotation

Regards, Sam

### LINKED ISSUE #2270: how to include our own model in the model list
Hello, Can someone share the process/steps to include our own like yolov3, yolov5, or faster rcnn model into the **Model** list for automatic/ semi automatic annotation. I wanna add my own model which i have trained on my own dataset.

### LINKED ISSUE #2926: Custom model for automatic annotation
Hello, I have seen many issues asking this but I am not sure about the answer. 

I have some data and I want to train my own model to annotate. Is it possible to do that? If so, are there instructions somewhere on how to do that? 

Also is there a specific framework required for training i.e. pytorch, tensorflow?

Thanks

### PR DESCRIPTION
**Pull request by nmanovic**
Fixed #2926
Fixed #2270
Fixed #2259
Fixed #2185

### Motivation and context

Users and customers have problems with running automatic and semi-automatic annotation in CVAT using serverless function. They don't know how to integrate own DL models. The tutorial answers on many typical questions which I saw as github issues or requests in gitter.

### How has this been tested?

- Clone CVAT repo and CD to cvat
- Checkout the branch `nm/serverless_tutorial`
- Go to `site` directory
- Run `hugo server -D` command
- Open `http://localhost:1313/docs/manual/advanced/serverless-tutorial/` page

### Checklist

- [x] I submit my changes into the `develop` branch
- [x] I have added description of my changes into [CHANGELOG](https://github.com/opencv/cvat/blob/develop/CHANGELOG.md) file
- [x] I have updated the [documentation](
  https://github.com/opencv/cvat/blob/develop/README.md#documentation) accordingly
- [ ] I have added tests to cover my changes
- [x] I have linked related issues ([read github docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [ ] I have increased versions of npm packages if it is necessary ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
[cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning), [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2021 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

### CODE REVIEW

**Comment by nmanovic** [comment]
@TOsmanov , @aschernov , could you please help to review the documentation and find any problems?

**Comment by aschernov** [comment]
@nmanovic , sure.

File: serverless/pytorch/facebookresearch/detectron2/retinanet/nuclio/main.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+from detectron2.data.datasets.builtin_meta import COCO_CATEGORIES
+
+CONFIG_FILE = "detectron2/configs/COCO-Detection/retinanet_R_101_FPN_3x.yaml"
+CONFIG_OPTS = ["MODEL.WEIGHTS", "model_final_971ab9.pkl"]
+CONFIDENCE_THRESHOLD = 0.5
+
+def init_context(context):
+    context.logger.info("Init context...  0%")
+    if torch.cuda.is_available():
+        functionconfig = yaml.safe_load(open("/opt/nuclio/function-gpu.yaml"))
```
Comment (IRDonch) [inline serverless/pytorch/facebookresearch/detectron2/retinanet/nuclio/main.py:20]:
```suggestion
        with open("/opt/nuclio/function-gpu.yaml", 'rb') as function_file:
            functionconfig = yaml.safe_load(function_file)
```

File: serverless/pytorch/facebookresearch/detectron2/retinanet/nuclio/main.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    cfg = get_cfg()
+    cfg.merge_from_file(CONFIG_FILE)
+    cfg.merge_from_list(CONFIG_OPTS)
+    cfg.MODEL.RETINANET.SCORE_THRESH_TEST = CONFIDENCE_THRESHOLD
+    cfg.MODEL.ROI_HEADS.SCORE_THRESH_TEST = CONFIDENCE_THRESHOLD
+    cfg.MODEL.PANOPTIC_FPN.COMBINE.INSTANCES_CONFIDENCE_THRESH = CONFIDENCE_THRESHOLD
+    cfg.freeze()
+    predictor = DefaultPredictor(cfg)
+
+    setattr(context.user_data, 'model_handler', predictor)
```
Comment (IRDonch) [inline serverless/pytorch/facebookresearch/detectron2/retinanet/nuclio/main.py:35]:
```suggestion
    context.user_data.model_handler = predictor
```

File: serverless/pytorch/facebookresearch/detectron2/retinanet/nuclio/main.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    labels_spec = functionconfig['metadata']['annotations']['spec']
+    labels = {item['id']: item['name'] for item in json.loads(labels_spec)}
+    setattr(context.user_data, "labels", labels)
+
+    context.logger.info("Init context...100%")
+
+def handler(context, event):
+    context.logger.info("Run retinanet-R101 model")
+    data = event.body
+    buf = io.BytesIO(base64.b64decode(data["image"].encode('utf-8')))
```
Comment (IRDonch) [inline serverless/pytorch/facebookresearch/detectron2/retinanet/nuclio/main.py:45]:
```suggestion
    buf = io.BytesIO(base64.b64decode(data["image"]))
```

It should be unnecessary to encode the string, as `b64decode` accepts character strings.

File: serverless/pytorch/facebookresearch/detectron2/retinanet/nuclio/main.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    cfg.MODEL.RETINANET.SCORE_THRESH_TEST = CONFIDENCE_THRESHOLD
+    cfg.MODEL.ROI_HEADS.SCORE_THRESH_TEST = CONFIDENCE_THRESHOLD
+    cfg.MODEL.PANOPTIC_FPN.COMBINE.INSTANCES_CONFIDENCE_THRESH = CONFIDENCE_THRESHOLD
+    cfg.freeze()
+    predictor = DefaultPredictor(cfg)
+
+    setattr(context.user_data, 'model_handler', predictor)
+    labels_spec = functionconfig['metadata']['annotations']['spec']
+    labels = {item['id']: item['name'] for item in json.loads(labels_spec)}
+    setattr(context.user_data, "labels", labels)
```
Comment (IRDonch) [inline serverless/pytorch/facebookresearch/detectron2/retinanet/nuclio/main.py:38]: This element of `user_data` never seems to actually get used in the handler - you're using `COCO_CATEGORIES` instead.

File: serverless/tensorflow/faster_rcnn_inception_v2_coco/nuclio/function-gpu.yaml [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        - kind: RUN
+          value: apt install curl
+        - kind: WORKDIR
+          value: /opt/nuclio
+
+      postCopy:
+        - kind: RUN
+          value: curl -O http://download.tensorflow.org/models/object_detection/faster_rcnn_inception_v2_coco_2018_01_28.tar.gz
+        - kind: RUN
+          value: tar -xzf faster_rcnn_inception_v2_coco_2018_01_28.tar.gz && rm faster_rcnn_inception_v2_coco_2018_01_28.tar.gz
```
Comment (IRDonch) [inline serverless/tensorflow/faster_rcnn_inception_v2_coco/nuclio/function-gpu.yaml:113]: I would suggest to merge these layers into one; otherwise you end up with an intermediate layer that contains the tarball and takes up space for no benefit.

File: serverless/tensorflow/faster_rcnn_inception_v2_coco/nuclio/main.py [hunk 6]
```diff
@@ -11,7 +12,10 @@ def init_context(context):
     model_path = "/opt/nuclio/faster_rcnn/frozen_inference_graph.pb"
     model_handler = ModelLoader(model_path)
     setattr(context.user_data, 'model_handler', model_handler)
-    functionconfig = yaml.safe_load(open("/opt/nuclio/function.yaml"))
+    if tf.test.is_gpu_available() and tf.test.is_built_with_cuda():
```
Comment (IRDonch) [inline serverless/tensorflow/faster_rcnn_inception_v2_coco/nuclio/main.py:15]: The TensorFlow documentation says that `is_gpu_available` is deprecated in favor of `tf.config.list_physical_devices('GPU')`. Also, the `is_built_with_cuda` check seems redundant - a GPU can't be available if there's no CUDA support.

File: site/content/en/docs/contributing/rest-api-design.md [hunk 7]
```diff
@@ -1,8 +1,13 @@
-# REST API design principles
+---
```
Comment (IRDonch) [inline site/content/en/docs/contributing/rest-api-design.md:1]: The changes in this file seem unrelated to the PR topic.

File: serverless/tensorflow/faster_rcnn_inception_v2_coco/nuclio/function-gpu.yaml [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+        { "id":80, "name": "toaster" },
+        { "id":81, "name": "sink" },
+        { "id":83, "name": "refrigerator" },
+        { "id":84, "name": "book" },
+        { "id":85, "name": "clock" },
+        { "id":86, "name": "vase" },
+        { "id":87, "name": "scissors" },
+        { "id":88, "name": "teddy_bear" },
+        { "id":89, "name": "hair_drier" },
+        { "id":90, "name": "toothbrush" }
```
Comment (IRDonch) [inline serverless/tensorflow/faster_rcnn_inception_v2_coco/nuclio/function-gpu.yaml:89]: It seems redundant to include this long list of labels into both files. Maybe move it to a separate file, e.g. `labels.yml`?

**Review by IRDonch — COMMENTED** [review]
I've reviewed the code so far; will look at the tutorial itself next.

File: serverless/pytorch/facebookresearch/detectron2/retinanet/nuclio/function-gpu.yaml [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+        - kind: RUN
+          value: apt-get update && apt-get -y install curl git python3 python3-pip
+        - kind: WORKDIR
+          value: /opt/nuclio
+        - kind: RUN
+          value: pip3 install torch==1.8.1+cu111 torchvision==0.9.1+cu111 torchaudio==0.8.1 -f https://download.pytorch.org/whl/torch_stable.html
+        - kind: RUN
+          value: git clone https://github.com/facebookresearch/detectron2
+        - kind: RUN
+          value: pip3 install -e detectron2
```
[the remark this answers is not in the document]
Comment (nmanovic) [reply serverless/pytorch/facebookresearch/detectron2/retinanet/nuclio/function-gpu.yaml]: @IRDonch , experimentally understand that the advice will not work. I need a configuration file from the repo.

File: serverless/tensorflow/faster_rcnn_inception_v2_coco/nuclio/main.py (hunk 6 above)
Comment (nmanovic) [reply serverless/tensorflow/faster_rcnn_inception_v2_coco/nuclio/main.py]: I understand that I don't need the `if` at all.

File: serverless/tensorflow/faster_rcnn_inception_v2_coco/nuclio/function-gpu.yaml (hunk 8 above)
Comment (nmanovic) [reply serverless/tensorflow/faster_rcnn_inception_v2_coco/nuclio/function-gpu.yaml:89]: Probably it will not work. I need to have the information inside function*.yaml. Server requests labels directory from nuclio. I have to say that function.yaml and function-gpu.yaml are mostly the same in many cases. But I prefer to keep them both for simplicity now.

Comment (IRDonch) [reply serverless/tensorflow/faster_rcnn_inception_v2_coco/nuclio/function-gpu.yaml:89]: OK, got it.

File: serverless/pytorch/facebookresearch/detectron2/retinanet/nuclio/function-gpu.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (IRDonch) [reply serverless/pytorch/facebookresearch/detectron2/retinanet/nuclio/function-gpu.yaml]: It looks like you can load this configuration file by using [detectron2.model_zoo.get_config](https://detectron2.readthedocs.io/en/latest/modules/model_zoo.html?highlight=get_config#detectron2.model_zoo.get_config). Just replace `get_cfg()` with `get_config('COCO-Detection/retinanet_R_101_FPN_3x.yaml')`. This will load the internal copy of the config file (which is installed in the Python environment), so you'll no longer need to keep the repository in the working directory.

**Comment by TOsmanov** [comment]
@nmanovic 
the tutorial does not have enough information on installing `nuctl` on windows 10, `administration/advanced/installation_automatic_annotation.md` contain commands only for linux

File: serverless/pytorch/facebookresearch/detectron2/retinanet/nuclio/function-gpu.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (nmanovic) [reply serverless/pytorch/facebookresearch/detectron2/retinanet/nuclio/function-gpu.yaml]: Great finding. Thanks!

**Comment by sevaraMS** [comment]
When it will be ready???. I need tutorial

**Comment by sevaraMS** [comment]
I tried use onepanel, but it is not suitable for windows

**Comment by sevaraMS** [comment]
with openvino not worked))))))))))))))))

File: site/content/en/docs/contributing/rest-api-design.md (hunk 7 above)
Comment (nmanovic) [reply site/content/en/docs/contributing/rest-api-design.md:1]: Agree, but it is only one small file. Let's keep it in PR for simplicity.

**Comment by sevaraMS** [comment]
Do we have another way to add our own trained model for auto annotation????

**Comment by TOsmanov** [comment]
I encountered a problem when running `nuctl` from windows 10, I managed to deployment models only using ubuntu. I took the following actions:

1. Installation [ubuntu 18.04 from Microsoft store](https://www.microsoft.com/en-us/p/ubuntu-1804-lts/9n9tngvndl3q)

1. In settings `docker desktop` in tab `resourcrces` `wsl integration` enable integration for ubuntu-18.04
    
    image

1. Install `curl` command:
    ```sh
    sudo apt install curl
    ```
1. Installation `nuctl` on ubuntu, with use command from [manuals on nuclio CLI](https://github.com/nuclio/nuclio/blob/development/docs/reference/nuctl/nuctl.md):
    ```sh
    curl -s https://api.github.com/repos/nuclio/nuclio/releases/latest \
                | grep -i "browser_download_url.*nuctl.*$(uname)" \
                | cut -d : -f 2,3 \
                | tr -d \" \
                | wget -O nuctl -qi - && chmod +x nuctl
    ```
 1. Installation `git` on ubuntu:
    ```sh
    sudo apt install git
    ```
 1. Cloning to the repository and moving:
    ```
    git clone https://github.com/openvinotoolkit/cvat.git --branch nm/serverless_tutorial
    cd cvat
    ```
1. Executing commands from serverless tutorial

There was a problem with the deployment of models 
(similar to this #2828) I solved it by adding it to the `function.yaml` file of the each model, attribute `port: 32001`

```
triggers:
myHttpTrigger:
    maxWorkers: 1
    kind: 'http'
    workerAvailabilityTimeoutMilliseconds: 10000
    attributes:
    port: 32001 <----- add this ( different ports for different models)
    maxRequestBodySize: 33554432 # 32MB
```



May be worth adding information that Windows 10 users are better off deploying models through ubuntu, or adding this information to section "Troubleshooting"?

**Comment by nmanovic** [comment]
> I encountered a problem when running `nuctl` from windows 10, I managed to deployment models only using ubuntu. I took the following actions:
> 
> 1. Installation [ubuntu 18.04 from Microsoft store](https://www.microsoft.com/en-us/p/ubuntu-1804-lts/9n9tngvndl3q)
> 2. In settings `docker desktop` in tab `resourcrces` `wsl integration` enable integration for ubuntu-18.04
>    image
> 3. Install `curl` command:
>    ```shell
>    sudo apt install curl
>    ```
> 4. Installation `nuctl` on ubuntu, with use command from [manuals on nuclio CLI](https://github.com/nuclio/nuclio/blob/development/docs/reference/nuctl/nuctl.md):
>    ```shell
>    curl -s https://api.github.com/repos/nuclio/nuclio/releases/latest \
>                | grep -i "browser_download_url.*nuctl.*$(uname)" \
>                | cut -d : -f 2,3 \
>                | tr -d \" \
>                | wget -O nuctl -qi - && chmod +x nuctl
>    ```
> 5. Installation `git` on ubuntu:
>    ```shell
>    sudo apt install git
>    ```
> 6. Cloning to the repository and moving:
>    ```
>    git clone https://github.com/openvinotoolkit/cvat.git --branch nm/serverless_tutorial
>    cd cvat
>    ```
> 7. Executing commands from serverless tutorial
> 
> There was a problem with the deployment of models
> (similar to this #2828) I solved it by adding it to the `function.yaml` file of the each model, attribute `port: 32001`
> 
> ```
> triggers:
> myHttpTrigger:
>     maxWorkers: 1
>     kind: 'http'
>     workerAvailabilityTimeoutMilliseconds: 10000
>     attributes:
>     port: 32001 <----- add this ( different ports for different models)
>     maxRequestBodySize: 33554432 # 32MB
> ```
> 
> May be worth adding information that Windows 10 users are better off deploying models through ubuntu, or adding this information to section "Troubleshooting"?

Thanks for these instructions. I will add the information into the tutorial.

**Comment by nmanovic** [comment]
@dmitriikhurtin , could you please help me to understand the issue? In the PR I see a couple of errors from pylint linter. For example,

```json
    {
        "type": "error",
        "module": "main",
        "obj": "init_context",
        "line": 19,
        "column": 12,
        "path": "serverless/openvino/omz/intel/semantic-segmentation-adas-0001/nuclio/main.py",
        "symbol": "too-many-function-args",
        "message": "Too many positional arguments for constructor call",
        "message-id": "E1121"
    }
```


If I check the file locally, the linter doesn't complain. How should I run the linter locally to reproduce the problem?

```console
$ echo "Pylint version: "`pylint --version | head -1`
Pylint version: pylint 2.7.0
$ echo "The files will be checked: "`echo ${changed_files_pylint}`
The files will be checked: /home/nmanovic/Workspace/cvat/serverless/openvino/omz/intel/semantic-segmentation-adas-0001/nuclio/main.py
$ pylint --verbose ${changed_files_pylint} --output-format json
Using config file /home/nmanovic/Workspace/cvat/.pylintrc
[]
```
